### ISSUE TITLE
CC0016 should not copy the variable with readonly members

### ISSUE THREAD

**Issue by Haacked** [code-cracker/code-cracker#645]
Suppose I have the following class:

```csharp
public class HttpHandlerRouteHandler<THttpHandler> : IRouteHandler where THttpHandler : IHttpHandler
{
    readonly Func<RequestContext, THttpHandler> _handlerFactory;

    public HttpHandlerRouteHandler(Func<RequestContext, THttpHandler> handlerFactory)
    {
        if (handlerFactory == null) throw new ArgumentNullException(nameof(handlerFactory));
        _handlerFactory = handlerFactory;
    }

    public IHttpHandler GetHttpHandler(RequestContext requestContext)
    {
        return _handlerFactory(requestContext);
    }
}
```

Note that `_handlerFactory` is a `readonly` field so it can't generally be changed on more than one thread. In this particular case, it also can't be null. So there is no need to copy the event handler to a variable before firing it.

But Code Cracker still recommends I copy it. Not a big deal since I'm betting this is tricky to get right. But thought I'd mention it. :smile:

Summarizing:

If the variable **is readonly** and it is being assigned on constructor or directly on the field declaration, do not raise `CC0016` anymore.

We will create another issue to raise an error when a readonly reference field that is never assigned is used for anything.

@Vossekop is working on it.

**Comment by carloscds** [comment]
@Haacked Thanks for your comment about this!

**Comment by giggio** [comment]
@Haacked On the `GetHttpHandler` method there are only two options for `_handlerFactory`: null or not null, and, as you said, that cannot be changed. Not a single thread, but at all, as being readonly only allows it's value to change during object construction. So, there is no way that the value is going to change, and therefore, there is no need to check.

A better approach would be to simply `_handlerFactory?.Invoke`:

````csharp
public IHttpHandler GetHttpHandler(RequestContext requestContext)
{
    return _handlerFactory?.Invoke(requestContext);
}
````

But that would fail unless we declare that `THttpHandler` is a `class`:

````
public class HttpHandlerRouteHandler<THttpHandler> : IRouteHandler where THttpHandler : class, IHttpHandler
````

So, yes, the copying is unnecessary, but the check for null is not. And we cannot reliably fall back to `?.`.

So, I am suggesting that we simply remove the copy when the field is readonly.

Also, there is another bug to fix. I applied the fix, and it removed the return statement, ending up like this:

````csharp
public IHttpHandler GetHttpHandler(RequestContext requestContext)
{
    var handler = _handlerFactory;
    if (handler != null)
        handler(requestContext);
}
````

This also has to be fixed. I will update the description of this issue with the expected output.

**Comment by Haacked** [comment]
> On the GetHttpHandler method there are only two options for _handlerFactory: null or not null

Well, in this case it cannot be null because the constructor for `HttpHandlerRouteHandler` would throw if it were null. I'm not sure how difficult it is to do that sort of flow analysis with Roslyn or not. For example, 
if I did a `Debug.Assert(_handlerFactory != null);` just before calling `_handlerFactory(...)` would Code Cracker still tell me I need to use `.?`?

**Comment by giggio** [comment]
@Haacked We can totally check for that. We are already doing that for `CC0031` (see it on [UseInvokeMethodToFireEventTests.cs](https://github.com/code-cracker/code-cracker/blob/3619b07ded508eaa6c91215692e9d2ead622a883/test/CSharp/CodeCracker.Test/Design/UseInvokeMethodToFireEventTests.cs#L241-L248)), which is similar to this one. I have updated the description again to reflect the discussion.
Also, I will create another issue to take care of the scenario of when a `readonly` reference field is used, as it will always be null.

**Comment by giggio** [comment]
Created issue #647 to track Readonly reference field is never assigned and is used.

**Comment by Vossekop** [comment]
@code-cracker/owners I'd like to try my hand at fixing this.
@giggio by updating the description I take it you mean the original post by @Haacked, and that the part under "Summarizing:" is what needs to happen, no?

**Comment by giggio** [comment]
@Vossekop Go for it.
Yes, under the `summarizing` I tried to summarize what we discussed in this issue comments.

**Comment by Vossekop** [comment]
@giggio since #647 is now closed as willnotfix, how should I handle situations where the initializer sets the read only variable to null. I had first assumed this should be ignored by CC0016 and picked up by whatever initializer should be made in #647, however since that one got axed I am unsure what should be done. It is clearly something that we might want to detect, but i feel CC0016's message is not applicable to the situation (also the suggested fix is somewhat illogical).

Also, i presume its ok to fix some spelling/grammar errors in the message/description for the analyzer.

**Comment by Vossekop** [comment]
Ok, so I've thought about this some more and have come to the conclusion that the ideal solution would have the following:

The warning stays for read only members if it can not be determined that they are initialized for certain to a value other than null. Meaning that if we encounter an initializer that sets the member to null then we will show the warning. If the constructor contains syntax that could possibly bypass an initializer in some cases (like an if statement) then the warning stays as well.

This is quite hard to get watertight (for instance what to do with an initializer in a for-each loop over a collection that might have no elements), but except for some corner cases I think I have a pretty good solution. Just have to document that sucker before doing a pull request.

**Comment by giggio** [comment]
@Vossekop Yes, you can fix spelling and grammar errors. :)
#647 was cut because C# already has a diagnostic for exactly that, `CS0649`.

Regarding the problem of not being sure the field was initialized, I would simply verify if the assignment is inside any conditional statement, like `if`, `for`, `foreach`, etc. Any of those means that the field could end up null, and `CC0016` still applies. Also, any return statements before the assignment, including the ones inside conditional statements, would also imply that the field could end up null, and `CC0016` still applies.

**Comment by Vossekop** [comment]
That is indeed what I did + if there is an assignment to null anywhere in the constructor CC0016 still applies. Also if not all constructors have an assignment then CC0016 still applies (if it was not assigned where it was defined obviously). Also if both if an else branch have an assignment then I don't apply CC0016, and also for switch statements where all cases assign a non null value.

**Comment by giggio** [comment]
Cool! Nice job!

**Comment by giggio** [comment]
Closed on #660.
